piet-coregraphics/src/ct_helpers.rs

```diff
@@ -0,0 +1,203 @@
+            let size = CTFramesetterSuggestFrameSizeWithConstraints(
+                self.0.as_concrete_TypeRef(),
+                range,
+                std::ptr::null(),
+                constraints,
+                &mut fit_range,
+            );
+            (size, fit_range)
+        }
+    }
+
+    pub(crate) fn create_frame(&self, range: CFRange, path: &CGPathRef) -> Frame {
+        Frame(self.0.create_frame(range, path))
+    }
+}
+
+impl Frame {
+    pub(crate) fn get_lines(&self) -> CFArray<CTLine> {
+        unsafe { TCFType::wrap_under_get_rule(CTFrameGetLines(self.0.as_concrete_TypeRef())) }
+    }
+
+    pub(crate) fn get_line(&self, line_number: usize) -> Option<CTLine> {
+        let idx: CFIndex = line_number.try_into().ok()?;
+        let lines = self.get_lines();
+        lines
+            .get(idx)
+            .map(|l| unsafe { TCFType::wrap_under_get_rule(l.as_concrete_TypeRef()) })
+    }
+
+    pub(crate) fn get_line_origins(&self, range: CFRange) -> Vec<CGPoint> {
+        let mut origins = vec![CGPoint::new(0.0, 0.0); range.length as usize];
+        unsafe {
+            CTFrameGetLineOrigins(self.0.as_concrete_TypeRef(), range, origins.as_mut_ptr());
+        }
+        origins
+    }
+}
+
+impl<'a> Line<'a> {
+    pub(crate) fn new(inner: &'a impl Deref<Target = CTLine>) -> Line<'a> {
+        Line(Cow::Borrowed(inner.deref()))
+    }
+
+    pub(crate) fn get_string_range(&self) -> CFRange {
+        unsafe { CTLineGetStringRange(self.0.as_concrete_TypeRef()) }
+    }
+
+    pub(crate) fn get_typographic_bounds(&self) -> TypographicBounds {
+        let mut out = TypographicBounds::default();
+        let width = unsafe {
+            CTLineGetTypographicBounds(
+                self.0.as_concrete_TypeRef(),
+                &mut out.ascent,
+                &mut out.descent,
+                &mut out.leading,
+            )
+        };
+        out.width = width;
+        out
+    }
+
+    pub(crate) fn get_string_index_for_position(&self, position: CGPoint) -> CFIndex {
+        unsafe { CTLineGetStringIndexForPosition(self.0.as_concrete_TypeRef(), position) }
+    }
+
+    /// return the 'primary' and 'secondary' offsets on the given line that the boundary of the
+    /// character at the provided index.
+    ///
+    /// I don't know what the secondary offset is for. There are docs at:
+    /// https://developer.apple.com/documentation/coretext/1509629-ctlinegetoffsetforstringindex
+    pub(crate) fn get_offset_for_string_index(&self, index: CFIndex) -> (CGFloat, CGFloat) {
+        let mut secondary: f64 = 0.0;
+        let primary = unsafe {
+            CTLineGetOffsetForStringIndex(self.0.as_concrete_TypeRef(), index, &mut secondary)
+        };
+        (primary, secondary)
+    }
+}
+
+impl<'a> From<CTLine> for Line<'a> {
+    fn from(src: CTLine) -> Line<'a> {
+        Line(Cow::Owned(src))
+    }
+}
+
+#[link(name = "CoreText", kind = "framework")]
+extern "C" {
```

Comment:
I was thinking not to, but on reflection I think it would be nice to make my use of these go through code review in those crates, so I'm going to upstream as much as I can.